plugins/network-elements/nuage-vsp/src/com/cloud/network/resource/NuageVspResourceConfiguration.java

```diff
@@ -0,0 +1,310 @@
+    }
+
+    public NuageVspResourceConfiguration cmsUserPassword(String cmsUserPassword) {
+        this._cmsUserPassword = cmsUserPassword;
+        return this;
+    }
+
+    public String port() {
+        return this._port;
+    }
+
+    public NuageVspResourceConfiguration port(String port) {
+        this._port = port;
+        return this;
+    }
+
+    public String apiVersion() {
+        return this._apiVersion;
+    }
+
+    public NuageVspResourceConfiguration apiVersion(String apiVersion) {
+        this._apiVersion = apiVersion;
+        return this;
+    }
+
+    public String apiRelativePath() {
+        return this._apiRelativePath;
+    }
+
+    public NuageVspResourceConfiguration apiRelativePath(String apiRelativePath) {
+        this._apiRelativePath = apiRelativePath;
+        return this;
+    }
+
+    public String retryCount() {
+        return this._retryCount;
+    }
+
+    public NuageVspResourceConfiguration retryCount(String retryCount) {
+        this._retryCount = retryCount;
+        return this;
+    }
+
+    public String retryInterval() {
+        return this._retryInterval;
+    }
+
+    public NuageVspResourceConfiguration retryInterval(String retryInterval) {
+        this._retryInterval = retryInterval;
+        return this;
+    }
+
+    public String nuageVspCmsId() {
+        return this._nuageVspCmsId;
+    }
+
+    public NuageVspResourceConfiguration nuageVspCmsId(String nuageVspCmsId) {
+        this._nuageVspCmsId = nuageVspCmsId;
+        return this;
+    }
+
+    public String getRootPath(){
+        return "https://" + _hostName + ":" + _port + "/nuage";
+    }
+
+    public String getApiPath() {
+            return "https://" + _hostName + ":" + _port + "/nuage/api/" + _apiVersion;
+    }
+
+    public NuageVspApiVersion getApiVersion() throws ConfigurationException {
+        try {
+            if(_apiVersion != null) {
+                return NuageVspApiVersion.fromString(_apiVersion);
+            }
+            return null;
+        } catch (IllegalArgumentException e) {
+            throw new ConfigurationException("Incorrect API version");
+        }
+    }
+
+    public Map<String, String> build() {
+        Map<String, String> map = new HashMap<>();
+        if (_guid != null) map.put(GUID, _guid);
+        if (_zoneId != null) map.put(ZONE_ID, _zoneId);
+        if (_hostName != null) map.put(HOST_NAME, _hostName);
+        if (_cmsUser != null) map.put(CMS_USER, _cmsUser);
+        if (_cmsUserPassword != null) map.put(CMS_USER_PASSWORD, _cmsUserPassword);
+        if (_port != null) map.put(PORT, _port);
+        if (_apiVersion != null) map.put(API_VERSION, _apiVersion);
+        if (_apiRelativePath != null) map.put(API_RELATIVE_PATH, _apiRelativePath);
+        if (_retryCount != null) map.put(RETRY_COUNT, _retryCount);
+        if (_retryInterval != null) map.put(RETRY_INTERVAL, _retryInterval);
+        if (_nuageVspCmsId != null) map.put(NUAGE_VSP_CMS_ID, _nuageVspCmsId);
```

Comment:
Please consider creating a utility method such as the following to DRY out this code:

```
public  static <K, V> void putWhenNotNull(final Map<K, V> map, K key, V value) {

     Preconditions.checkArgument(map != null);
     Preconditions.checkArgument(key != null);

     if (value != null) {
           map.put(key, value);
     }

}
```